src/rpp/rpp/operators/retry_when.hpp

```diff
@@ -0,0 +1,158 @@
+        RPP_NO_UNIQUE_ADDRESS TObservable       observable;
+        RPP_NO_UNIQUE_ADDRESS Notifier          notifier;
+
+        template<typename T>
+        void on_next(T&&) const
+        {
+            observable.subscribe(std::move(observer));
+        }
+
+        void on_error(const std::exception_ptr& err) const
+        {
+            observer.on_error(err);
+        }
+
+        void on_completed() const
+        {
+            observer.on_completed();
+        }
+
+        void set_upstream(const disposable_wrapper& d) { observer.set_upstream(d); }
+
+        bool is_disposed() const { return observer.is_disposed(); }
+    };
+
+    template<rpp::constraint::observer     TObserver,
+             rpp::constraint::observable   TObservable,
+             rpp::constraint::decayed_type Notifier>
+    struct retry_when_impl_strategy
+    {
+        using preferred_disposable_strategy = rpp::details::observers::none_disposable_strategy;
+
+        RPP_NO_UNIQUE_ADDRESS mutable TObserver   observer;
+        RPP_NO_UNIQUE_ADDRESS mutable TObservable observable;
+        RPP_NO_UNIQUE_ADDRESS mutable Notifier    notifier;
+
+        template<typename T>
+        void on_next(T&& v) const
+        {
+            observer.on_next(std::forward<T>(v));
+        }
+
+        void on_error(const std::exception_ptr& err) const
+        {
+            std::optional<std::invoke_result_t<Notifier, std::exception_ptr>> notifier_obs;
+            try
+            {
+                notifier_obs.emplace(notifier(err));
+            }
+            catch (...)
+            {
+                observer.on_error(std::current_exception());
+            }
+            if (notifier_obs.has_value())
+            {
+                std::move(notifier_obs).value().subscribe(retry_when_impl_inner_strategy<TObserver, TObservable, Notifier>{std::move(observer), std::move(observable), std::move(notifier)});
+            }
+        }
+
+        void on_completed() const
+        {
+            observer.on_completed();
+        }
+
+        void set_upstream(const disposable_wrapper& d) { observer.set_upstream(d); }
+
+        bool is_disposed() const { return observer.is_disposed(); }
+    };
+
+    template<rpp::constraint::observable TObservable, rpp::constraint::decayed_type Notifier>
+    struct retry_when_impl_t : lift_operator<retry_when_impl_t<TObservable, Notifier>, TObservable, Notifier>
+    {
+        using lift_operator<retry_when_impl_t<TObservable, Notifier>, TObservable, Notifier>::lift_operator;
+
+        template<rpp::constraint::decayed_type T>
+        struct operator_traits
+        {
+            using result_type = T;
+
+            template<rpp::constraint::observer_of_type<result_type> TObserver>
+            using observer_strategy = retry_when_impl_strategy<TObserver, TObservable, Notifier>;
+        };
+
+        template<rpp::details::observables::constraint::disposable_strategy Prev>
+        using updated_disposable_strategy = Prev;
```

Comment:
actually not, it can be infinite amount of disposables in case of infinite amount of retries

Reply:
let's make custom disposable instead like this (https://github.com/victimsnino/ReactivePlusPlus/blob/v2/src/rpp/rpp/operators/take_until.hpp#L38), pass it to original observer, accumulate all inner disposbles inside it and clear (https://github.com/victimsnino/ReactivePlusPlus/blob/b8b0ed589352bdb6bf05bb9b534ed0010b13cb40/src/rpp/rpp/sources/concat.hpp#L98) internals during moving from observable to callback and back. else it could grow internal container

Reply:
> actually not, it can be infinite amount of disposables in case of infinite amount of retries

Is there a way to amend the disposable test to cover this case?

Reply:
i think it is possible. 
https://github.com/victimsnino/ReactivePlusPlus/blob/v2/src/tests/utils/disposable_observable.hpp#L104 there is tested fixed_disposable strategy, so, you have to test_operator_over_observable_with_disposable such an callable which would reproduce such an situation)